[Docs] Add Migration guide + some API updates (#8716)
espressif/arduino-esp32: docs/source/migration_guide/2.x_to_3.0.rst

#########################
Migration from 2.x to 3.0
#########################

Introduction
------------

> This is a guide to highlight breaking changes in the API and to help the migration of projects from versions 2.X to version 3.0 of the Arduino ESP32 core. 

Reviewer: add space line between line 8 and 9.

> This is a guide to highlight breaking changes in the API and to help the migration of projects from versions 2.X to version 3.0 of the Arduino ESP32 core. 

Reviewer: Can we mentioned here that v2.X is based on ESP-IDF v4.4 and 3.0 is based on ESP-IDF 5.X?

> For more information about the changes and new features, check `RELEASE NOTES <https://github.com/espressif/arduino-esp32/releases>`_.

Reviewer: For more information about all changes and new features, check project `RELEASE NOTES <https://github.com/espressif/arduino-esp32/releases>`_.

> 

Reviewer: What about adding heading here, something like:

**Breaking Changes in APIs**

For more information about the new API, check :doc:`/api/ledc`.

Removed APIs
************

* ``ledcSetup``
* ``ledcAttachPin``

New APIs
********

* ``ledcAttach`` used to set up the LEDC pin (merged ``ledcSetup`` and ``ledcAttachPin`` functions).
* ``timerGetFrequency`` used to get the actual frequency of the timer.
* ``timerAttachInterruptArg`` used to attach the interrupt to a timer using arguments.

Changes in APIs
***************

* ``ledcDetachPin`` renamed to ``ledcDetach``.
* In all functions, input parameter ``channel`` has been changed to ``pin``.

RMT
---

For more information about the new API, check :doc:`/api/rmt`.

Removed APIs
************

* ``_rmtDumpStatus``
* ``rmtSetTick``
* ``rmtWriteBlocking``
* ``rmtEnd``
* ``rmtBeginReceive``
* ``rmtReadData``

New APIs
********

* ``rmtWriteAsync``
* ``rmtTransmitCompleted``
* ``rmtSetRxMinThreshold``

Changes in APIs
***************

* In all functions, input parameter ``rmt_obj_t* rmt`` has been changed to ``int pin``.
* ``rmtInit`` return parameter changed to bool.
* ``rmtInit`` input parameter ``bool tx_not_rx`` has been changed to ``rmt_ch_dir_t channel_direction``.

> * ``rmtInit`` new input parameter ``uint32_t frequency_Hz`` to set frequency of RMT channel as function ``rmtSetTick`` was removed.

Reviewer 2: ```suggestion
* ``rmtInit`` new input parameter ``uint32_t frequency_Hz`` to set frequency of RMT channel (as function ``rmtSetTick`` was removed).
```

> * ``rmtWrite`` now sending data in blocking mode - only returns after sending all data or by timeout. For Async mode use new ``rmtWriteAsync`` function.

Reviewer 2: ```suggestion
* ``rmtWrite`` now sending data in blocking mode. It only returns after sending all data or through a timeout. For Async mode use the new ``rmtWriteAsync`` function.
```

* ``rmtWrite`` new input parameter ``uint32_t timeout_ms``.
* ``rmtLoop`` renamed to ``rmtWriteLooping``.
* ``rmtRead`` input parameters changed to ``int pin, rmt_data_t* data, size_t *num_rmt_symbols, uint32_t timeout_ms``.
* ``rmtReadAsync`` input parameters changed to ``int pin, rmt_data_t* data, size_t *num_rmt_symbols``.
* ``rmtSetRxThreshold`` renamed to ``rmtSetRxMaxThreshold`` and input parameter ``uint32_t value`` has been changed to ``uint16_t idle_thres_ticks``.

> * ``rmtSetCarrier`` input parameters ``uint32_t low, uint32_t high`` has been changed to ``uint32_t frequency_Hz, float duty_percent``.

Reviewer 2: ```suggestion
* ``rmtSetCarrier`` input parameters ``uint32_t low, uint32_t high`` have been changed to ``uint32_t frequency_Hz, float duty_percent``.
```

Removed APIs
************

* ``timerGetConfig``
* ``timerSetConfig``
* ``timerSetDivider``
* ``timerSetCountUp``
* ``timerSetAutoReload``
* ``timerGetDivider``
* ``timerGetCountUp``
* ``timerGetAutoReload``
* ``timerAlarmEnable``
* ``timerAlarmDisable``
* ``timerAlarmWrite``
* ``timerAlarmEnabled``
* ``timerAlarmRead``
* ``timerAlarmReadMicros``
* ``timerAlarmReadSeconds``
* ``timerAttachInterruptFlag``

New APIs
********

* ``timerAlarm`` used to set up Alarm for the timer and enable it automatically (merged ``timerAlarmWrite`` and ``timerAlarmEnable`` functions).
* ``timerGetFrequency`` used to get the actual frequency of the timer.
* ``timerAttachInterruptArg`` used to attach the interrupt to a timer using arguments.

Changes in APIs
***************

* ``timerBegin`` has now only 1 parameter (frequency). There is an automatic calculation of the divider using different clock sources
  to achieve the selected frequency.
* ``timerAttachInterrupt`` has now only 2 parameters. The ``edge`` parameter has been removed.

UART
----

Removed APIs
************

> * ``uartDetachPins``

Reviewer 3: This function is internal and not available in HardwareSerial API.
This is internal to esp32-hal-uart.h only, but user don't use such APIs.

  
New APIs
********

* ``uart_get_RxPin`` used to get UART RX pin number.
* ``uart_get_TxPin`` used to get UART TX pin number.

> 

Reviewer 3: These are internal helper functions, not exposed in HardwareSerial class.

Changes in APIs
***************

* ``uartEnd`` input parameter ``uart_t* uart`` has been changed to ``uint8_t uart_num``.
* ``uartSetMode`` input parameter ``uint8_t mode`` has been changed to ``uart_mode_t mode``.

> * ``uartSetHwFlowCtrlMode`` input parameter ``uint8_t mode`` has been changed to ``uart_hw_flowcontrol_t mode``.

Reviewer 3: These functions are from esp32-hal-uart.c, which are internal only. 
Arduino has HardwareSerial class used by `Serial` which is the UART interface for users.
